app/src/main/java/com/mrkenhoff/qalculate/ui/CalculatorButton.kt

```diff
@@ -1,28 +1,98 @@
+class CalculatorButton : ConstraintLayout {
+    var onCenterAction: () -> Unit = {}
+    var onLongCenterAction: () -> Unit = {}
+    var onTopLeftAction: () -> Unit = {}
+    var onTopRightAction: () -> Unit = {}
+    var onBottomRightAction: () -> Unit = {}
+    var onBottomLeftAction: () -> Unit = {}
 
-    constructor(context: Context) : super(context) {
-        setup(null)
-    }
+    var text : String
+        get() = binding.centerButton.text.toString()
+        set(value) { binding.centerButton.text = value }
+
+    var topLeftText : String
+        get() = binding.topLeftText.text.toString()
+        set(value) { binding.topLeftText.text = value }
+
+    var topRightText : String
+        get() = binding.topRightText.text.toString()
+        set(value) { binding.topRightText.text = value }
+
+    var bottomRightText : String
+        get() = binding.bottomRightText.text.toString()
+        set(value) { binding.bottomRightText.text = value }
+
+    var bottomLeftText : String
+        get() = binding.bottomLeftText.text.toString()
+        set(value) { binding.bottomLeftText.text = value }
+
+    private var xStart = 0.0f
+    private var yStart = 0.0f
+
+    private var _binding: CalculatorButtonBinding? = null
+    private val binding get() = _binding!!
+
+    constructor(context: Context) : super(context) { }
+
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { }
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) { }
+
+    init {
+        _binding = CalculatorButtonBinding.inflate(LayoutInflater.from(context), this, true)
+
+        binding.centerButton.setOnClickListener { onCenterAction() }
+        binding.centerButton.setOnLongClickListener {
+            onLongCenterAction()
+            true
+        }
 
-    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
-        setup(attrs)
+        binding.centerButton.setOnTouchListener(View.OnTouchListener { view, motionEvent ->
+            when (motionEvent.action){
+                MotionEvent.ACTION_DOWN, MotionEvent.ACTION_POINTER_DOWN -> {
+                    onTouchDown(motionEvent.x, motionEvent.y)
+                }
+                MotionEvent.ACTION_UP -> {
+                    onTouchUp(motionEvent.x, motionEvent.y)
+                }
+            }
+            return@OnTouchListener false
+        })
     }
 
-    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
-        setup(attrs)
+    private fun onTouchDown(x: Float, y: Float) {
+        xStart = x
+        yStart = y
     }
 
-    private fun setup(attrs: AttributeSet?) {
-        setOnClickListener {
-            EventBus.getDefault().post(ButtonEvent(typedText ?: text.toString()))
+    private fun onTouchUp(x: Float, y: Float) {
+        val dist = sqrt(x*x + y * y)
+        if (dist > 8) { // TODO: Do not hardcode distance
```

Comment:
Would be great to have some visual feedback when that distance is reached. For example, the action_move event could make one of the texts larger/bold when releasing the finger at that position will trigger the corresponding action.

Reply:
:+1: Exactly what I thought, but did not have time this weekend to implement it :)